lib/model/user_group.dart

```diff
@@ -55,8 +71,34 @@ class UserGroupStoreImpl extends PerAccountStoreBase with UserGroupStore {
     return _groups.values;
   }
 
+  @override
+  bool selfInGroupSetting(GroupSettingValue value) {
+    return switch (value) {
+      GroupSettingValueNamed() =>
+        _selfInGroup(value.groupId),
+      GroupSettingValueNameless() =>
+        value.directMembers.contains(selfUserId)
+          || value.directSubgroups.any(_selfInGroup),
+    };
+  }
+
+  bool _selfInGroup(int groupId) {
+    final group = _groups[groupId];
+    if (group == null) return false; // TODO(log); should know all groups
+    // TODO(perf), TODO(#814): memoize which groups the self-user is in,
+    //   to save doing this depth-first search on each permission check
+    return group.members.contains(selfUserId)
+      || group.directSubgroupIds.any(_selfInGroup);
```

Comment:
I have a draft branch for this optimization. But it's a bit tricky to get right, so I've separated it out in order to let this PR focus on all the other aspects of the semantics of these permissions.